core/corehttp/gateway_handler.go

```diff
@@ -443,13 +445,12 @@ func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request) {
 
 	var newPath string
 	if len(rsegs) > 1 {
-		newPath = path.Join(rsegs[2:])
+		newPath = coreapi.JoinComponents(rsegs[2:])
 	}
 
 	var newcid *cid.Cid
-	rnode, err := core.Resolve(ctx, i.node.Namesys, i.node.Resolver, rootPath)
-	switch ev := err.(type) {
-	case path.ErrNoLink:
+	_, rnode, err := i.api.ResolveNode(ctx, rootPath)
+	if err == coreiface.ErrNotFound {
```

Comment:
When changing \core.Resolve` to `api.ResolveNode`, there is an extra `api.node.DAG.Get` call hiding behind. Its result is discarded. Could it be possible to reuse it for later?